state/modelsummaries.go

```diff
@@ -0,0 +1,435 @@
+	invalidLocalUsers set.Strings
+
+	// incompleteUUIDs are ones that are missing some information, we should treat them as not being available
+	// we wait to strip them out until we're done doing all the processing steps.
+	incompleteUUIDs set.Strings
+}
+
+func newProcessorFromModelDocs(st *State, modelDocs []modelDoc, user names.UserTag, isSuperuser bool) *modelSummaryProcessor {
+	p := &modelSummaryProcessor{
+		st:          st,
+		user:        user,
+		isSuperuser: isSuperuser,
+	}
+	p.summaries = make([]ModelSummary, len(modelDocs))
+	p.indexByUUID = make(map[string]int, len(modelDocs))
+	p.modelUUIDs = make([]string, len(modelDocs))
+	for i, doc := range modelDocs {
+		var cloudCred string
+		if names.IsValidCloudCredential(doc.CloudCredential) {
+			cloudCred = names.NewCloudCredentialTag(doc.CloudCredential).String()
+		}
+		p.summaries[i] = ModelSummary{
+			Name:               doc.Name,
+			UUID:               doc.UUID,
+			Life:               doc.Life,
+			Owner:              doc.Owner,
+			ControllerUUID:     doc.ControllerUUID,
+			SLALevel:           string(doc.SLA.Level),
+			SLAOwner:           doc.SLA.Owner,
+			CloudTag:           names.NewCloudTag(doc.Cloud).String(),
+			CloudRegion:        doc.CloudRegion,
+			CloudCredentialTag: cloudCred,
+			/// Users:              make(map[string]UserAccessInfo),
+			/// Machines:           make(map[string]MachineModelInfo),
+		}
+		p.indexByUUID[doc.UUID] = i
+		p.modelUUIDs[i] = doc.UUID
+	}
+	return p
+}
+
+func (p *modelSummaryProcessor) fillInFromConfig() error {
+	// We use the raw settings because we are reading across model UUIDs
+	rawSettings, closer := p.st.database.GetRawCollection(settingsC)
+	defer closer()
+
+	remaining := set.NewStrings(p.modelUUIDs...)
+	settingIds := make([]string, len(p.modelUUIDs))
+	for i, uuid := range p.modelUUIDs {
+		settingIds[i] = uuid + ":" + modelGlobalKey
+	}
+	query := rawSettings.Find(bson.M{"_id": bson.M{"$in": settingIds}})
+	var doc settingsDoc
+	iter := query.Iter()
+	for iter.Next(&doc) {
+		idx, ok := p.indexByUUID[doc.ModelUUID]
+		if !ok {
+			// How could it return a doc that we don't have?
+			continue
+		}
+		remaining.Remove(doc.ModelUUID)
+
+		cfg, err := config.New(config.NoDefaults, doc.Settings)
+		if err != nil {
+			// err on one model should kill all the other ones?
+			return errors.Trace(err)
+		}
+		detail := &(p.summaries[idx])
+		detail.ProviderType = cfg.Type()
+		detail.DefaultSeries = config.PreferredSeries(cfg)
+		if agentVersion, exists := cfg.AgentVersion(); exists {
+			detail.AgentVersion = &agentVersion
+		}
+	}
+	if err := iter.Close(); err != nil {
+		return errors.Trace(err)
+	}
+	if !remaining.IsEmpty() {
+		// XXX: What error is appropriate? Do we need to care about models that its ok to be missing?
+		return errors.Errorf("could not find settings/config for models: %v", remaining.SortedValues())
```

Comment:
I'd prefer to log this as a Warning and return nil (i.e. continue processing up the stack)